### PR TITLE
Add an outline of a Gossip server in Go.

### PR DESCRIPTION
**Pull request by AlCutter**
Outline of the HTTPS server functionality described in http://datatracker.ietf.org/doc/draft-linus-trans-gossip-ct/?include_text=1 (version 02)

Missing lots of functionality; no validation of anything AT ALL, no Trusted Auditor support, etc. etc., but it's a start.

### CODE REVIEW

File: go/gossip/handler_test.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+        ]
+      }`
+)
+
+func CreateAndOpenStorage() *Storage {
+	dir, err := ioutil.TempDir("", "handlertest")
+	if err != nil {
+		log.Fatalf("Failed to get temporary dir for test: %v", err)
+	}
+	*dbName = dir + "/gossip.db"
```
Comment (benlaurie) [inline go/gossip/handler_test.go]: Eww. Why is this not local to Storage?

File: go/gossip/storage.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+		return errors.New("attempting to call Open() on an already Open()'d Storage")
+	}
+	s.db, err = sql.Open("sqlite3", *dbName)
+	if err != nil {
+		return err
+	}
+	if _, err := s.db.Exec(schema); err != nil {
+		return err
+	}
+	if s.insertChain, err = s.db.Prepare(insertChain); err != nil {
```
Comment (benlaurie) [inline go/gossip/storage.go]: Man. Seems like their ought to be an easier way than all this repetition?

File: go/gossip/storage.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+	}
+	return id, nil
+}
+
+// insertThingOrSelectID will attempt to execute the insert Statement (under transaction tx), if that fails due to
+// a unique primary key constraint, it will look up that primary key by executing the getID Statement.
+// Returns the ID associated with persistent thing, or an error describing the failure.
+func insertThingOrSelectID(tx *sql.Tx, insert *sql.Stmt, getID *sql.Stmt, thing interface{}) (int64, error) {
+	txInsert := tx.Stmt(insert)
+	txGetID := tx.Stmt(getID)
```
Comment (benlaurie) [inline go/gossip/storage.go]: O.M.G.

You should use txInsert.LastInsertId().

Don't believe it needs to be in a transaction, either.

File: go/gossip/storage.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+	r, err := txInsert.Exec(thing)
+	if err != nil {
+		switch err.(type) {
+		case sqlite3.Error:
+			if err.(sqlite3.Error).Code == sqlite3.ErrConstraint {
```
Comment (benlaurie) [inline go/gossip/storage.go]: Can you explain this notation to me? :-)

File: go/gossip/storage.go (hunk 3 above)
Comment (benlaurie) [inline go/gossip/storage.go]: err. I meant r.LastInsertId(), below.

Comment (benlaurie) [inline go/gossip/storage.go]: Oh, I see I am completely confused. :-) Never mind.

File: go/gossip/storage.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+// AddSCTFeedback stores the passed in feedback object.
+func (s *Storage) AddSCTFeedback(feedback SCTFeedback) error {
+	tx, err := s.db.Begin()
+	if err != nil {
+		return err
+	}
+	for _, f := range feedback.Feedback {
+		chainID, err := s.addChainIfNotExists(tx, f.X509Chain)
+		if err != nil {
+			return err
```
Comment (benlaurie) [inline go/gossip/storage.go]: Leaves transaction open.

File: go/gossip/storage.go [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+		default:
+			return err
+		}
+	}
+	return nil
+}
+
+// AddSCTFeedback stores the passed in feedback object.
+func (s *Storage) AddSCTFeedback(feedback SCTFeedback) error {
+	tx, err := s.db.Begin()
```
Comment (benlaurie) [inline go/gossip/storage.go]: Why does this need to be in a transaction?

File: go/gossip/storage.go (hunk 2 above)
Comment (benlaurie) [inline go/gossip/storage.go]: One obvious possibility...

Most of these SQL strings could be synthesized on the fly (e.g "SELECT COUNT(*) FROM <table>"), you could keep a map from strings to prepared statements and prepare them on the fly.

File: go/gossip/handler_test.go (hunk 1 above)
Comment (AlCutter) [inline go/gossip/handler_test.go]: Fair point, is now.

File: go/gossip/storage.go (hunk 6 above)
Comment (AlCutter) [inline go/gossip/storage.go]: I guess it doesn't really need to be at the moment, but if we ever allowed the server to ditch old feedback/pollen then we'd potentially have a race here.

File: go/gossip/storage.go (hunk 5 above)
Comment (AlCutter) [inline go/gossip/storage.go]: I assumed that it'd `Rollback()` automatically when the `Tx` went out of scope, but you're right - it doesn't.
Fixed that up with some `defer` trickery.

File: go/gossip/storage.go (hunk 2 above)
Comment (AlCutter) [inline go/gossip/storage.go]: Thing is with that approach is that you lose the compile-time checking that you didn't misspell the key somewhere.

I've tried something else, have a look and see what you think.

File: go/gossip/storage.go (hunk 4 above)
Comment (AlCutter) [inline go/gossip/storage.go]: This is a "type assertion" (kinda like a `dynamic_cast` I guess; gives you back a reference to `err` which has the `sqlite3.Error` type, or `nil` if you guessed wrong), but I've realised that I can have it for free by capturing the value of the `err.(type)` above.

File: go/gossip/handler.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		rw.Header().Add("Allow", "POST")
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	decoder := json.NewDecoder(req.Body)
+	var feedback SCTFeedback
+	if err := decoder.Decode(&feedback); err != nil {
+		rw.WriteHeader(http.StatusBadRequest)
+		rw.Write([]byte(fmt.Sprintf("Invalid SCT Feedback received: %v", err)))
```
Comment (benlaurie) [inline go/gossip/handler.go]: This stanza occurs a lot ... refactor?

Comment (AlCutter) [inline go/gossip/handler.go]: Good call, done.

File: go/gossip/storage.go [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+			sctID, err := s.addSCTIfNotExists(tx, sct)
+			if err != nil {
+				return err
+			}
+			if err = s.addSCTFeedbackIfNotExists(tx, chainID, sctID); err != nil {
+				return err
+			}
+		}
+	}
+	return tx.Commit()
```
Comment (benlaurie) [inline go/gossip/storage.go]: First assert err == nil? Or put the commit in the defer also?

**Comment by benlaurie** [comment]
Other than that comment...

File: go/gossip/storage.go (hunk 8 above)
Comment (AlCutter) [inline go/gossip/storage.go]: Moved this (and the same pattern in the other method which used a TX) to the `defer` block.

**Pull request merged**